### PR TITLE
Fi 497 - Check must support elements for fixed values

### PR DESCRIPTION
**Pull request by czh-orz**
This PR was originally meant for slices marked as "must support". I've decided to go about this by checking for fixed values when checking must support elements and then removing duplicates. Before, when an element had a minimum cardinality of 0 and a fixed value, we never checked if the element found was the fixed value, and I'm guessing the profile validation wouldn't have picked it up because the minimum cardinality is 0.

**Submitter:**
- [X] This pull request describes why these changes were made
- [X] Internal ticket for this PR: https://oncprojectracking.healthit.gov/support/browse/FI-497
- [X] Internal ticket links to this PR
- [X] Internal ticket is properly labeled (Community/Program)
- [X] Internal ticket has a justification for its Community/Program label
- [X] Code diff has been reviewed for extraneous/missing code
- [] Tests are included and test edge cases
- [X] Tests/code quality metrics have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: lib/modules/uscore_v3.1.0/bodyheight_sequence.rb [hunk 1]
```diff
@@ -406,45 +402,45 @@ def validate_resource_item(resource, property, value)
...
-          'Observation.component.valueSampledData',
-          'Observation.component.valueTime',
-          'Observation.component.valueDateTime',
-          'Observation.component.valuePeriod',
-          'Observation.component.dataAbsentReason'
+          { path: 'Observation.status', fixed_value: '' },
+          { path: 'Observation.category', fixed_value: '' },
+          { path: 'Observation.category.coding', fixed_value: '' },
+          { path: 'Observation.category.coding.system', fixed_value: 'http://terminology.hl7.org/CodeSystem/observation-category' },
+          { path: 'Observation.category.coding.code', fixed_value: 'vital-signs' },
```
Comment (Jammjammjamm) [inline lib/modules/uscore_v3.1.0/bodyheight_sequence.rb]: I think as currently written this PR will pass an Observation if it contains the code in one category and the system in another rather than requiring them to both be present in the same slice.

File: generator/uscore/uscore_generator.rb [hunk 2]
```diff
@@ -349,8 +349,8 @@ def create_must_support_test(sequence)
           )
         }
 
-        must_support_elements = sequence[:must_supports].select { |must_support| must_support[:type] == 'element' }
-        must_support_elements.each do |element|
+        sequence[:must_supports] = sequence[:must_supports].uniq
```
Comment (Jammjammjamm) [inline generator/uscore/uscore_generator.rb]: Are both this `uniq` and the `delete_if` in the metadata extractor necessary?

File: lib/modules/uscore_v3.1.0/bp_sequence.rb [hunk 3]
```diff
@@ -630,157 +482,85 @@ def validate_resource_item(resource, property, value)
...
+          { path: 'Observation.component.valueBoolean.code', fixed_value: '' },
+          { path: 'Observation.component.valueInteger.code', fixed_value: '' },
+          { path: 'Observation.component.valueRange.code', fixed_value: '' },
+          { path: 'Observation.component.valueRatio.code', fixed_value: '' },
+          { path: 'Observation.component.valueSampledData.code', fixed_value: '' },
+          { path: 'Observation.component.valueTime.code', fixed_value: '' },
+          { path: 'Observation.component.valueDateTime.code', fixed_value: '' },
+          { path: 'Observation.component.valuePeriod.code', fixed_value: '' },
+          { path: 'Observation.component', fixed_value: '' },
+          { path: 'Observation.component.code', fixed_value: '' },
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/modules/uscore_v3.1.0/bp_sequence.rb]: This is kind of in the grey area. So component.code is must supported, but component.code.coding isn't. And component.code.coding has a slice with a fixed code and system. I think the code and system part of the validation should be left to the profile validation.

File: lib/modules/uscore_v3.1.0/bp_sequence.rb [hunk 4]
```diff
@@ -630,157 +482,85 @@ def validate_resource_item(resource, property, value)
...
+          { path: 'Observation.component.valueInteger.value', fixed_value: '' },
+          { path: 'Observation.component.valueRange.value', fixed_value: '' },
+          { path: 'Observation.component.valueRatio.value', fixed_value: '' },
+          { path: 'Observation.component.valueSampledData.value', fixed_value: '' },
+          { path: 'Observation.component.valueTime.value', fixed_value: '' },
+          { path: 'Observation.component.valueDateTime.value', fixed_value: '' },
+          { path: 'Observation.component.valuePeriod.value', fixed_value: '' },
+          { path: 'Observation.component.valueQuantity.unit', fixed_value: '' },
+          { path: 'Observation.component.valueCodeableConcept.unit', fixed_value: '' },
+          { path: 'Observation.component.valueString.unit', fixed_value: '' },
```
Comment (Jammjammjamm) [inline lib/modules/uscore_v3.1.0/bp_sequence.rb]: These `value`, `unit`, `system`, and `code` must supports are only for the required slices in this profile which limit `value[x]` to `valueQuantity`.

File: lib/modules/uscore_v3.1.0/bodyheight_sequence.rb [hunk 5]
```diff
@@ -406,45 +402,45 @@ def validate_resource_item(resource, property, value)
...
-          'Observation.component.valueBoolean',
-          'Observation.component.valueInteger',
-          'Observation.component.valueRange',
-          'Observation.component.valueRatio',
-          'Observation.component.valueSampledData',
-          'Observation.component.valueTime',
-          'Observation.component.valueDateTime',
-          'Observation.component.valuePeriod',
-          'Observation.component.dataAbsentReason'
+          { path: 'Observation.status', fixed_value: '' },
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/modules/uscore_v3.1.0/bodyheight_sequence.rb]: Fixed

File: lib/modules/uscore_v3.1.0/bp_sequence.rb [hunk 6]
```diff
@@ -630,157 +482,85 @@ def validate_resource_item(resource, property, value)
...
+          { path: 'Observation.component.valueCodeableConcept', fixed_value: '' },
+          { path: 'Observation.component.valueString', fixed_value: '' },
+          { path: 'Observation.component.valueBoolean', fixed_value: '' },
+          { path: 'Observation.component.valueInteger', fixed_value: '' },
+          { path: 'Observation.component.valueRange', fixed_value: '' },
+          { path: 'Observation.component.valueRatio', fixed_value: '' },
+          { path: 'Observation.component.valueSampledData', fixed_value: '' },
+          { path: 'Observation.component.valueTime', fixed_value: '' },
+          { path: 'Observation.component.valueDateTime', fixed_value: '' },
+          { path: 'Observation.component.valuePeriod', fixed_value: '' },
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/modules/uscore_v3.1.0/bp_sequence.rb]: Made it so it only checks for Observation.component.value so it will pass if any value[x] is found.

File: lib/modules/uscore_v3.1.0/bp_sequence.rb (hunk 4 above)
Comment (czh-orz) [reply lib/modules/uscore_v3.1.0/bp_sequence.rb]: Fixed by checking for any value[x]. I think profile validation would be able to check that it's valueQuantity specifically.

**Comment by arscan** [comment]
Have all comments been addressed?

**Comment by czh-orz** [comment]
> Have all comments been addressed?

No, I'm trying to redo how the slice validation works.

File: lib/modules/uscore_v3.1.0/bp_sequence.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply lib/modules/uscore_v3.1.0/bp_sequence.rb]: Didn't we decide to remove the checks for component completely since it doesn't really make sense for most of the vital signs?

**Comment by Jammjammjamm** [comment]
I encountered this in the smoking status sequence:
Screen Shot 2020-01-30 at 10 45 54 AM

and this in the Practitioner sequence:
Screen Shot 2020-01-30 at 10 47 24 AM

**Comment by czh-orz** [comment]
> I encountered this in the smoking status sequence:
> Screen Shot 2020-01-30 at 10 45 54 AM
> 
> and this in the Practitioner sequence:
> Screen Shot 2020-01-30 at 10 47 24 AM

Should be fixed now.

**Comment by Jammjammjamm** [comment]
I still have the problem in the Practitoner sequence:
```
INFO | Inferno | Starting Test: USCPR-09 [All must support elements are provided in the Practitioner resources returned.]
ERROR | Inferno | Fatal Error: undefined method `values' for #<Array:0x00007fd81e630a08>
Did you mean?  values_at
ERROR | Inferno | [
  "inferno/lib/modules/uscore_v3.1.0/us_core_practitioner_sequence.rb:297:in `block (2 levels) in <class:USCore310PractitionerSequence>'",
  "inferno/lib/modules/uscore_v3.1.0/us_core_practitioner_sequence.rb:295:in `reject'",
  "inferno/lib/modules/uscore_v3.1.0/us_core_practitioner_sequence.rb:295:in `block in <class:USCore310PractitionerSequence>'",
  "inferno/lib/app/sequence_base.rb:427:in `instance_eval'",
  "inferno/lib/app/sequence_base.rb:427:in `run_test'"...]
```

**Comment by czh-orz** [comment]
> I still have the problem in the Practitoner sequence:
> 
> ```
> INFO | Inferno | Starting Test: USCPR-09 [All must support elements are provided in the Practitioner resources returned.]
> ERROR | Inferno | Fatal Error: undefined method `values' for #<Array:0x00007fd81e630a08>
> Did you mean?  values_at
> ERROR | Inferno | [
>   "inferno/lib/modules/uscore_v3.1.0/us_core_practitioner_sequence.rb:297:in `block (2 levels) in <class:USCore310PractitionerSequence>'",
>   "inferno/lib/modules/uscore_v3.1.0/us_core_practitioner_sequence.rb:295:in `reject'",
>   "inferno/lib/modules/uscore_v3.1.0/us_core_practitioner_sequence.rb:295:in `block in <class:USCore310PractitionerSequence>'",
>   "inferno/lib/app/sequence_base.rb:427:in `instance_eval'",
>   "inferno/lib/app/sequence_base.rb:427:in `run_test'"...]
> ```

Woops, missed that part. Actually resolved now.
